File: strategy/c-sharp/ShoppingCart.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Strategy
{
    // UML diyagramındaki Context sınıfına denk gelmektedir.
    // UML diyagramındaki Strategy(IPayment) arayüzüne ait referansı tutmaktadır.
    class ShoppingCart
    {
        private List<Item> _items;
        private IPayment _payment;

        public ShoppingCart()
        {
            _items = new List<Item>();
        }

        public void SetPaymentMethod(IPayment payment)
        {
            _payment = payment;
        }

        public void AddItem(Item item)
        {
            _items.Add(item);
        }

        public int CalculateTotal()
        {
            int sum = 0;
            foreach (Item item in _items)
            {
                sum += item.GetPrice();
            }
            return sum;
        }

        public void Pay()
        {
            int amount = CalculateTotal();
            _payment.Pay(amount);
        }
    }
}

File: strategy/go/shopping-cart.go
package main

// UML diyagramındaki Context sınıfına denk gelmektedir.
// UML diyagramındaki Strategy(IPayment) arayüzüne ait referansı tutmaktadır.
type shoppingCart struct {
	items   []*Item
	payment Payment
}

func NewShoppingCart() *shoppingCart {
	return &shoppingCart{
		items: make([]*Item, 0),
	}
}

func (s *shoppingCart) AddItem(item *Item) {
	s.items = append(s.items, item)
}

func (s *shoppingCart) SetPaymentMethod(p Payment) {
	s.payment = p
}

func (s shoppingCart) GetTotal() float64 {
	total := 0.0
	for _, item := range s.items {
		total += item.GetPrice()
	}
	return total
}

func (s shoppingCart) Pay() {
	s.payment.Pay(s.GetTotal())
}
